libflux/src/semantic/types.rs

```diff
@@ -959,8 +959,71 @@ impl MaxTvar for Function {
 }
 
 impl Function {
-    fn unify(self, _: Self, _: &mut TvarKinds) -> Result<Substitution, Error> {
-        unimplemented!();
+    // In order to unify two function types, it is necessary and sufficient to unify their argument and return types.
+    // This seems straight forward however there are some subtleties when unifying argument types.
+    fn unify(
+        self,
+        other: Self,
+        cons: &mut TvarKinds,
+        fresh: &mut Fresher,
+    ) -> Result<Substitution, Error> {
+        let f = self;
+        let mut g = other;
+        // We start with a validation phase from steps 1 and 3:
+        //  1. Verify that a required argument of f is in the required/optional arguments of g (and viceversa).
+        //  3. If a pipe arg was used then the other type must specify a pipe arg too.
+        // This is done at the beginning both to avoid starting unifying when there is no need to,
+        // and in order to error using f and g, otherwise, the borrow checker would not make us
+        // error while we have borrowed some of their internal pieces (while iterating for example).
```

Comment:
If this is not possible and we have to clone, I'm ok with that. We can tackle performance once we have correctness.